invenio_records_resources/references/entity_resolvers/base.py

```diff
@@ -19,13 +19,19 @@ def _parse_ref_dict(reference_dict, strict=True):
     The ``strict`` parameter controls if the number of keys in the
     reference dict is checked strictly or not.
     """
-    keys = list(reference_dict.keys())
+    try:
+        keys = list(reference_dict.keys())
 
-    if strict and len(keys) != 1:
-        raise ValueError(
-            "Reference dicts may only have one property! "
-            f"Offending dict: {reference_dict}"
-        )
+        if strict and len(keys) != 1:
+            raise ValueError(
+                "Reference dicts may only have one property! "
+                f"Offending dict: {reference_dict}"
+            )
+
+    except AttributeError:
+        # if reference dict is not really a dict
+        # happens if entity does not match the type and gets passed to _parse_ref_dict)
+        keys = None
```

Comment:
minor: I'm not sure about the exception handling here. Wouldn't it make sense to instead check if the `reference_dict` is actually a `dict` when initializing the proxy?

Reply:
It's about these lines here:

https://github.com/inveniosoftware/invenio-records-resources/blob/ec55cf197ddfc5a0231818345dbc8822573b2201/invenio_records_resources/references/registry.py#L83-L91

This works for classes which inherit from `Record`, since this is also a dict. Then only a `ValueError` is raised. If another class is passed here (e.g. `ResultItem`), then an `AttributeError` will be raised.

So it could be ok to just add the `AttributeError` to the except block, so this is caught as well.

Reply:
Ok I see now. We should only aim at calling `.reference_entity(...)` on API-class/Record instances and not `ResultItem` instances. I'm a bit worried about `AttributeError` masking e.g. an actual bug in the underlying code, in which case we don't have a way to differentiate.

I also suspect that the `ValueError` here is aimed at catching the exception raised at https://github.com/inveniosoftware/invenio-records-resources/blob/58503e2a6fdc7bd110f7d3bf3a51e7ca05fd63f0/invenio_records_resources/references/entity_resolvers/base.py#L167 which is raised/controlled by us.

Reply:
Do you mean `reference_entity(...)` or `matches_reference_dict(...)` should only be called on API-class/Record instances?
If second, a type check of the entity could work
```py
elif isinstance(entity, dict) and resolver.matches_reference_dict(entity):
  return entity```

So probably what you wrote in your first comment? :)

Reply:
Updated it to add this check